src/java.base/share/classes/jdk/internal/constant/MethodTypeDescImpl.java

```diff
@@ -105,18 +105,46 @@ public static MethodTypeDescImpl ofValidated(ClassDesc returnType, ClassDesc...
      * @jvms 4.3.3 Method Descriptors
      */
     public static MethodTypeDescImpl ofDescriptor(String descriptor) {
-        // Implicit null-check of descriptor
-        List<ClassDesc> ptypes = ConstantUtils.parseMethodDescriptor(descriptor);
-        int args = ptypes.size() - 1;
-        ClassDesc[] paramTypes = args > 0
-                ? ptypes.subList(1, args + 1).toArray(ConstantUtils.EMPTY_CLASSDESC)
-                : ConstantUtils.EMPTY_CLASSDESC;
-
-        MethodTypeDescImpl result = ofValidated(ptypes.get(0), paramTypes);
+        int rightBracket = descriptor.indexOf(')');
+        int len = descriptor.length() - rightBracket - 1;
+        if (rightBracket <= 0 || descriptor.charAt(0) != '(' || len == 0
+                || len != ConstantUtils.skipOverFieldSignature(descriptor, rightBracket + 1, descriptor.length(), true)
+        ) {
+            badMethodDescriptor(descriptor);
+        }
+        var returnType = ConstantUtils.resolveClassDesc(descriptor, rightBracket + 1, len);
+        var paramTypes = paramTypes(descriptor, 1, rightBracket);
+        var result = new MethodTypeDescImpl(returnType, paramTypes);
         result.cachedDescriptorString = descriptor;
         return result;
     }
 
+    private static ClassDesc[] paramTypes(String descriptor, int start, int end) {
+        int paramCount = 0;
+        for (int cur = start; cur < end; ) {
+            int len = ConstantUtils.skipOverFieldSignature(descriptor, cur, end, false);
```

Comment:
Allocating a buffer will slow down, so I used a long, with each 8 bits storing a parameter length, which can improve performance when the number of parameters is <= 8.

Reply:
Since we can't have empty descriptors (`len == 0`) you could leave a 0 in the bits to signify that the parameter is too large to fit and then re-parse it in the subsequent loop below. Might simplify the control flow a bit (no need for `largeParm`, maybe possible to merge the 2nd and 3rd loops)